File: Data Structures & Algorithms/three-integer-sum/submission-11.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        l = len(nums)

        res = []
      
        for i in range(l):
            n1 = nums[i]
            # Because no pos + pos + pos = 0. 
            if n1 > 0:
                break

            # If the last one equals this one. We continue.
            # Cause it'd create a duplicate. 
            # if i > 0 and n1 == nums[i - 1]:
            #     continue
                
            left = i + 1
            right = l - 1

            s = None
            while left < right:
                s = n1 + nums[left] + nums[right]
                lst = [n1, nums[left], nums[right]]
                if s > 0:
                    right -= 1
                elif s < 0:
                    left += 1
                elif lst not in res:
                    res.append(lst)
                    # We already used these. So we move in. 
                    left += 1
                    right -= 1

                    # We move left so we don't use duplicate L values. Cause that'd be a duplicate. 
                    # while nums[left] == nums[left - 1] and left < right:
                    #     left += 1
                else:
                    left += 1
                    right -= 1
                    

                    

        return res

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-11.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left < right) {
            final int mid = left + (right - left) / 2;

            if (nums[mid] >= nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        int pivot = left;

        int result = binarySearch(nums, 0, pivot, target);

        return result == -1 ? binarySearch(nums, pivot, nums.length - 1, target) : result;
    }

    public int binarySearch(int[] nums, int left, int right, int target) {
        while (left <= right) {
            final int mid = left + (right - left) / 2;
            final int value = nums[mid];

            if (value == target) 
                return mid;

            if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return -1;
    }
}
